### PR TITLE
js codegen: better error and peer-dependencies

### LINKED ISSUE #8101: Give better error message if multiple versions of @daml/* are loaded
A user on the forum ran into an error because their SDK versions in `daml.yaml` and `package.json` didn't match. As a result, the generated code pulled in a different version of `@daml/types` to the one pulled in by `package.json` resulting in an error related to not being able to look up the template. https://discuss.daml.com/t/i-get-trying-to-look-up-template-error-in-front-end/1715/3

Maybe the message "Trying to look up template" could be made more informative, but more importantly, I think it would be nice to emit an error (or a warning) on loading the library that more than one version is around and that that will usually lead to problems.

### PR DESCRIPTION
**Pull request by robin-da**
We add the daml-types/daml-ledger libraries as peer-dependencies instead
of dependencies. This will assure that the library is provided by the
importing library and avoid the situation when two different versions of
these libraries are imported by the generated and the user code.

This fixes #8101.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by robin-da** [comment]
@hurryabit It would be really nice to have a static check that yields a compiler error when two different versions of `@daml/types` are present in the dependency tree. I tried to put an interface in the global namespace with a different typed field for each library version but this happily compiles and the compiler seems to just pick one version when importing the two packages. Any ideas?

**Comment by garyverhaegen-da** [comment]
> @hurryabit It would be really nice to have a static check that yields a compiler error when two different versions of `@daml/types` are present in the dependency tree. I tried to put an interface in the global namespace with a different typed field for each library version but this happily compiles and the compiler seems to just pick one version when importing the two packages. Any ideas?

If you can access a common namespace, it seems like something along the lines of:
```
global_daml_types_version = (global_daml_types_version || []).append(this_version)
global_daml_types_version.forEach((v) => if (v !=== this_version) { exit loudly with an error });
```
should work, with js being single-threaded. Could be a boolean if we're confident we never load the same package twice, too.

**Comment by cocreature** [comment]
Afaict, this only works under the assumption that the generated code is compatible with multiple versions of `@daml/types`. I’m not sure that’s a valid assumption to make and it’s definitely not one we test for. If that assumption is broken, you will just end up with whatever runtime or maybe type errors come out of the mismatch.

I’d much rather try to detect the mismatch even if maybe via a somewhat crude method than try to support this. For a type error version, I would have tried to inject something in `Template`, either just a number or maybe something like the contract id branding we do atm.

**Comment by garyverhaegen-da** [comment]
So if `@daml/types` has a bit of code along the lines of:
```
if (global_daml_types_version) {
  exit loudly
} else {
  global_daml_types_version = this_version;
}
```
and every generated file starts with
```
if (global_daml_type_version !== this_version) { exit loudly; }
```
shouldn't that take care of this issue? I mean, admittedly it's not a static check, and it's not terribly elegant, but a crash a load time with a nice error message seems preferable to subtle behaviour-changing bugs.

Granted, load time can be any time in js, but I would argue for most well-structured programs loading is done at startup.

**Comment by robin-da** [comment]
I also rather have a compile time check. The template id branding we do already covers the runtime check. @cocreature could you elaborate on the "inject something into Template" id and how it is different from the branding we for contract ids?

**Comment by cocreature** [comment]
For `ContractId`s we have https://github.com/digital-asset/daml/blob/94578bd2fda020c5dc09fefabe144e428f8fa02c/language-support/ts/daml-types/index.ts#L296 to make sure that things don’t unify between templates. I obviously haven’t tried this but maybe we can inject that into the `Template` interface to make sure that two versions of that inferface don’t unify if they come from different library versions.

That said, Gary’s suggestion also sounds reasonable and while a typecheck is nicer, if it blows up immediately, this is probably good enough.

**Comment by robin-da** [comment]
@cocreature that's what I tried with interface, but this doesn't work. I see if I can make a runtime check work.

**Comment by cocreature** [comment]
@robin-da Do you have a branch of what you tried? Even if we don’t do it now, it might be useful to look into this later.

**Comment by robin-da** [comment]
@cocreature I added 
```
export type UniqueVersionType = '0.0.0-SDKVERSION';
declare global {
  interface DifferentLibraryVersionsLoaded {damlTypes: UniqueVersionType}
}``` 
to @daml/types.

**Comment by cocreature** [comment]
My suggestion was to modify the `Template` interface rather than adding a new interface.

**Comment by robin-da** [comment]
Yes, but it works the same way. It shouldn't matter whether you define an interface just for the check or reuse an existing one. Or am I missing something?

**Comment by robin-da** [comment]
They should be unified because they live in the same global namespace and have the same name.

**Comment by robin-da** [comment]
@cocreature added a `sdkVersion` field to the template interface and it has the targeted effect. I'll also give the global namespace another shot, maybe I was just testing it not right. But I think the `sdkVersion` in the template interface is nicer so this should go in.

File: language-support/ts/daml-ledger/index.test.ts [hunk 1]
```diff
@@ -72,6 +72,7 @@ jest.mock('isomorphic-ws', () => class {
 
 
 const Foo: Template<Foo, string, "foo-id"> = {
+  sdkVersion: '0.0.0-SDKVERSION',
```
Comment (cocreature) [inline language-support/ts/daml-ledger/index.test.ts:75]: This is going to break on release builds where the version is not `0.0.0`. Try setting `DAML_SDK_RELEASE_VERSION` to test this.

File: language-support/ts/daml-types/index.ts [hunk 2]
```diff
@@ -31,6 +31,10 @@ export interface Serializable<T> {
  */
 export interface Template<T extends object, K = unknown, I extends string = string> extends Serializable<T> {
   templateId: I;
+  /**
+   * @internal
+   */
+  sdkVersion: '0.0.0-SDKVERSION';
```
Comment (cocreature) [inline language-support/ts/daml-types/index.ts:37]: I’m confused, this is hardcoded not the actual SDK version, how does this work?

Comment (robin-da) [reply language-support/ts/daml-types/index.ts:37]: when we build the package with our bazel rules, all strings '0.0.0-SDKVERSION' in the source are replaced by the actual sdk version.

Comment (cocreature) [reply language-support/ts/daml-types/index.ts:37]: Have you checked this? My understanding was that we only replace things in `package.json` not the source files. Looking at the Bazel rules seems to confirm that.

Comment (cocreature) [reply language-support/ts/daml-types/index.ts:37]: ah nvm, we replace in all files.

**Review by cocreature — APPROVED** [review]
Thanks! LGTM

File: language-support/ts/daml-types/index.ts (hunk 2 above)
Comment (robin-da) [reply language-support/ts/daml-types/index.ts:37]: this should also resolve the above issue in the test, right?

Comment (cocreature) [reply language-support/ts/daml-types/index.ts:37]: Not quite sure, we do this in `pkg_npm` not during the build. I think the tests might still be affected. Easy enough to test with `DAML_SDK_RELEASE_VERSION`.

File: language-support/ts/daml-ledger/index.test.ts (hunk 1 above)
Comment (robin-da) [reply language-support/ts/daml-ledger/index.test.ts:75]: Looks like this works fine. Tested with different `DAML_SDK_RELEASE_VERSION` and `bazel run //language-support/ts/daml-ledger:test`

**Pull request merged**